src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -465,67 +645,170 @@ void postponeConditionallyUsedValues ()
+		if (instruction.hasSideEffect())
 		{
 			return null;
 		}
+		final int writeMask = instruction.operation().writesHiddenVariablesMask;
+		if ((writeMask & readMaskLaterInBlock) != 0
+			|| (writeMask & writeMaskLaterInBlock) != 0)
+		{
+			// Either the instruction writes something that a later instruction
+			// reads, or there are multiple writes that should maintain their
+			// order in case a later block has a dependent read.  Don't move it.
+			return null;
+		}
+		final int readMask = instruction.operation().readsHiddenVariablesMask;
+		if ((readMask & writeMaskLaterInBlock) != 0)
+		{
+			// The instruction reads something that a later instruction would
+			// clobber.  Don't move it.
+			return null;
+		}
+
 		final List<L2Register> written = instruction.destinationRegisters();
+		final List<L2SemanticValue> writtenSemanticValues = new ArrayList<>();
+		instruction.writeOperands().forEach(
+			write -> writtenSemanticValues.addAll(write.semanticValues()));
 		assert !written.isEmpty()
 			: "Every instruction should either have side effects or write to "
 			+ "at least one register";
-		if (!disjoint(written, registersConsumedLaterInSameBlock))
+		if (!disjoint(written, registersConsumedLaterInBlock))
 		{
 			// A later instruction in the current basic block consumes one of
-			// the values produced by the given instruction, so we can't move
+			// the registers produced by the given instruction, so we can't move
 			// the given instruction into later blocks.
 			return null;
 		}
-		final L2BasicBlock block = instruction.basicBlock;
-		if (block.successorEdgesCount() == 1)
+		if (!disjoint(
+			writtenSemanticValues, semanticValuesConsumedLaterInBlock))
+		{
+			// A later instruction in the current basic block consumes one of
+			// the semantic values produced by the given instruction, so we
+			// can't move the given instruction into later blocks.
+			return null;
+		}
+		if (candidateTargetEdges.size() == 1)
 		{
 			// There's only one successor edge.  Since the CFG is in edge-split
 			// form, the successor might have multiple predecessors.  Don't move
 			// across the edge in that case, since it may cause the instruction
 			// to run in situations that it doesn't need to.  When code
 			// splitting is eventually implemented, it should clean up this case
 			// by duplicating the successor block just for this edge.
-			final L2BasicBlock successor =
-				block.successorEdgeAt(0).targetBlock();
+			final L2PcOperand successorEdge =
+				candidateTargetEdges.iterator().next();
+			final L2BasicBlock successor = successorEdge.targetBlock();
 			if (successor.predecessorEdgesCount() > 1)
 			{
+				// One last chance to postpone this instruction.  Some
+				// idempotent operations prefer to be replicated in this
+				// situation, to increase the chance that they can migrate
+				// somewhere that their cost is lower to compute.  See
+				// L2_VIRTUAL_REIFY.
+				if (false) //TODO enable this when it works.
```

Comment:
So the commented out code is completely busted? Can you describe the problem?

Reply:
I removed it.  The problem it was trying to solve was that instructions were getting stuck during postponement by bumping into control flow merges, so it was allowing some kinds of instructions to be moved into the merged flow, even though some paths might reach the instruction after going through a duplicate.  Not a great solution.

One way to solve this is to combine "equivalent" instructions arriving at a control flow merge, modulo phi renames.  This is basically undoing the duplication of the original instruction.  I call this the "merge" approach.

Another way to solve it is to examine the graph and figure out where all the earliest uses are (uses that have no earlier uses along any histories), and duplicate the instructions as close to those uses as possible (there may be additional constraints based on the visibility of the instruction's needed semantic values and registers).  This approach I call "tunnel", since it can transport the instructions through arbitrarily shaped subgraphs to where they're needed.

This will take more thought, and will have to be solved another day.